feat: Add CREATE CREDENTIALS docs (#172)
GlareDB/glaredb.github.io: _glaredb/sql-commands/create-credentials.md

---
layout: default
title: CREATE CREDENTIALS
parent: SQL commands
---

# CREATE CREDENTIALS

Create credentials for accessing AWS and Google Cloud resources.

## Syntax

```sql
CREATE CREDENTIALS <credential-name>
    PROVIDER <provider>
    OPTIONS (<provider-options>);
```

| Field              | Description                                                       |
| ------------------ | ----------------------------------------------------------------- |
| `credential-name`  | Name for the credentials.                                         |

> | `provider`         | The cloud provider for these credentials are for. `aws` or `gcp`. |

Reviewer: ```suggestion
| `provider`         | The cloud provider these credentials are for. `aws` or `gcp`. |
```

I'd do something like Valid values: [`aws`, `gcp`].

| `provider-options` | Provider specific options. See below sections.                    |

## AWS credentials

AWS credentials allows GlareDB to read and write objects in S3.

Creating AWS credentials requires `access_key_id` and `secret_access_key`. These
correspond to an IAM user with permissions for accessing objects in S3.

```sql

> CREATE CREDENTIALS my_aws_creds

Author: This would be useful when running locally, but could be confusing when running with hybrid exec. It _would_ be interesting if we could forward `.aws/config` credentials automatically though...

    PROVIDER aws
    OPTIONS (
        access_key_id = 'my_access_key_id',
        secret_access_key = 'my_secret_access_key',

>     );

Author: No, only access key and secret key. Region is tied to the bucket, but credentials can be applied to buckets in multiple regions.

```

> We can use the above credentials to access objects in S3:

Reviewer: ```suggestion
The above credentials can be used to access objects in S3:
```

> We can use the above credentials to access objects in S3:

Reviewer 2: I find `above` to be sort of the documentation equivalent of a raw pointer: easy to break, hard to detect except at runtime, sometimes. 

Better I think to describe the state: ("After configuring the credentials"

```sql
SELECT * FROM parquet_scan('s3://my_bucket/data/*.parquet', my_aws_creds, region => 'us-east-1');
```

> We can also use credentials to copy output of a query to S3:

Reviewer: ```suggestion
As another example, the credentials can be used to copy output of a query to S3:
```

```sql
COPY ( SELECT 5 AS a, 6 AS b )
    TO 's3://my_bucket/data/output.parquet'
    CREDENTIALS my_aws_creds

>     ( region 'us-east-1' );

Author: To be consistent with options when creating external tables/databases (those don't use equals). The `=>` that you see with the table functions are standard named argument syntax, but I'd like to see those made optional (or allow `=`) and have that unified the `COPY` and `CREATE EXTERNAL ...` options.

```

> Note that both of these example require specifying `region`. This is required
> when interacting with S3, and should be the AWS region where the bucket is
> located.

Reviewer: ```suggestion
Note that both of these example require specifying `region`. Region is required
when interacting with S3, and always refers to the AWS region where the bucket is
located.
```

## GCP credentials

> GCP credentials allows GlareDB to readn and write objects in GCS.

Reviewer: ```suggestion
GCP credentials allows GlareDB to read and write objects in GCS.
```

Creating GCP credentials requires `service_account_key` to be provided as an
option. This should be a JSON encoded key for a service account with access to

> the GCS buckets you'll want to query.

Reviewer: Something like:

The `service_account_key` option is required when creating GCP credentials. `service_account_key` is
a JSON-encoded key for a service account. Only buckets that this service account has read permissions
for can be queryed.

```sql
CREATE CREDENTIALS my_gcp_creds
    PROVIDER gcp
    OPTIONS (
        service_account_key = 'my_gcp_service_account_key',
    );
```

> We can use the above credentials to access objects in GCS:

Reviewer: ```suggestion
The above credentials can be used to access objects in GCS:
```

```sql
SELECT * FROM parquet_scan('gs://my_bucket/data/*.parquet', my_gcp_creds);
```

> We can also use credentials to copy output of a query to GCS:

Reviewer: ```suggestion
As another example, the credentials can be used to copy output of a query to GCS:
```
